ext/nnstreamer/extra/nnstreamer_protobuf_grpc.cc

```diff
@@ -0,0 +1,371 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+/**
+ * GStreamer / NNStreamer gRPC/protobuf support for tensor src/sink
+ * Copyright (C) 2020 Dongju Chae <dongju.chae@samsung.com>
+ */
+/**
+ * @file    nnstreamer_protobuf_grpc.cc
+ * @date    21 Oct 2020
+ * @brief   gRPC/Protobuf wrappers for nnstreamer
+ * @see     https://github.com/nnstreamer/nnstreamer
+ * @author  Dongju Chae <dongju.chae@samsung.com>
+ * @bug     No known bugs except for NYI items
+ */
+
+#include "nnstreamer_protobuf_grpc.h"
+#include "nnstreamer.grpc.pb.h" /* Generated by `protoc` */
+
+#include <nnstreamer_log.h>
+
+#include <thread>
+
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/channel.h>
+#include <grpcpp/client_context.h>
+#include <grpcpp/create_channel.h>
+#include <grpcpp/security/credentials.h>
+#include <grpcpp/health_check_service_interface.h>
+
+#include <gst/base/gstdataqueue.h>
+
+using nnstreamer::protobuf::TensorService;
+using nnstreamer::protobuf::Tensors;
+using nnstreamer::protobuf::Tensor;
+
+using google::protobuf::Empty;
+
+namespace grpc {
+
+/**
+ * @brief NNStreamer gRPC service impl.
+ */
+class NNStreamerRPC final : public TensorService::Service {
+  public:
+    /** @brief constructor */
+    NNStreamerRPC (gboolean is_server, const gchar *host, const gint port):
+      is_server_ (is_server), host_ (host), port_ (port),
+      server_ (nullptr), stub_ (nullptr),
+      cb_ (nullptr), cb_data_ (nullptr), queue_ (nullptr)
+    {
+      gst_tensors_config_init (&config_);
+    }
+
+    /** @brief destructor */
+    ~NNStreamerRPC () {
+      if (server_.get ())
+        server_->Shutdown ();
+
+      if (worker_.joinable ())
```

Comment:
Note: In my case, join () in the destructor () is never passed. (https://github.com/nnstreamer/nnstreamer-ros/pull/56/files#diff-e85c333c6471ef6fc92208770d33712586b9c5974a98113adfad5f1213a40ae1R173) Please check this out.